packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp

```diff
@@ -358,8 +359,75 @@ initialize ()
     (! G->isFillComplete (), std::runtime_error, "If you call this method, "
      "the matrix's graph must be fill complete.  It is not.");
 
+  if (A_crs_->isUpperTriangular() && htsImpl_.is_null()) {
+    // Reverse the storage for an upper triangular matrix
+    using local_matrix_type = typename crs_matrix_type::local_matrix_type;
+
+    auto Alocal = A_crs_->getLocalMatrix();
+    auto ptr    = Alocal.graph.row_map;
+    auto ind    = Alocal.graph.entries;
+    auto val    = Alocal.values;
+
+    auto numRows = Alocal.numRows(), numCols = Alocal.numCols(), numNnz = Alocal.nnz();
+
+    typename decltype(ptr)::non_const_type  newptr ("ptr", ptr.dimension_0 ());
+    typename decltype(ind)::non_const_type  newind ("ind", ind.dimension_0 ());
+    decltype(val)                           newval ("val", val.dimension_0 ());
+
+    newptr(0) = 0;
+    for (local_ordinal_type row = 0, rowStart = 0; row < numRows; ++row) {
+      auto A_r = Alocal.row(numRows-1 - row);
+
+      auto numEnt = A_r.length;
+      for (local_ordinal_type k = 0; k < numEnt; ++k) {
+        newind(rowStart + k) = numCols-1 - A_r.colidx(numEnt-1 - k);
+        newval(rowStart + k) = A_r.value (numEnt-1 - k);
+      }
+      rowStart += numEnt;
+      newptr(row+1) = rowStart;
+    }
+
+    // Reverse maps
+    using map_type = typename crs_matrix_type::map_type;
+    Teuchos::RCP<map_type> newRowMap, newColMap;
+    {
+      // Reverse row map
+      auto rowMap = A_->getRowMap();
+      auto numElems = rowMap->getNodeNumElements();
+      auto rowElems = rowMap->getNodeElementList();
+
+      Teuchos::Array<global_ordinal_type> newRowElems(rowElems.size());
+      for (size_t i = 0; i < numElems; i++)
+        newRowElems[i] = rowElems[numElems-1 - i];
+
+      newRowMap = Teuchos::rcp(new map_type(rowMap->getGlobalNumElements(), newRowElems, rowMap->getIndexBase(), rowMap->getComm()));
+    }
+    {
+      // Reverse column map
+      auto colMap = A_->getColMap();
+      auto numElems = colMap->getNodeNumElements();
+      auto colElems = colMap->getNodeElementList();
+
+      Teuchos::Array<global_ordinal_type> newColElems(colElems.size());
+      for (size_t i = 0; i < numElems; i++)
+        newColElems[i] = colElems[numElems-1 - i];
+
+      newColMap = Teuchos::rcp(new map_type(colMap->getGlobalNumElements(), newColElems, colMap->getIndexBase(), colMap->getComm()));
+    }
+
+    // Construct new matrix
+    local_matrix_type newLocalMatrix("Upermuted", numRows, numCols, numNnz, newval, newptr, newind);
+
+    A_crs_ = Teuchos::rcp(new crs_matrix_type(newRowMap, newColMap, A_crs_->getDomainMap(), A_crs_->getRangeMap(), newLocalMatrix));
+
+    isInternallyChanged_ = true;
+  }
+
   if (Teuchos::nonnull (htsImpl_))
+  {
     htsImpl_->initialize (*A_crs_);
```

Comment:
@mhoemmen, if ShyLU/HTS is enabled, then Belos and units tests in Ifpack2 for HTS get enabled, too.